File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/servers/library/items/media/files/properties/uri/MediaFileUriProvider.java
package com.lasthopesoftware.bluewater.servers.library.items.media.files.properties.uri;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.net.Uri;
import android.provider.MediaStore;
import android.support.v4.content.ContextCompat;
import android.support.v4.content.LocalBroadcastManager;

import com.lasthopesoftware.bluewater.servers.connection.IConnectionProvider;
import com.lasthopesoftware.bluewater.servers.library.items.media.files.IFile;
import com.lasthopesoftware.bluewater.servers.library.items.media.files.properties.CachedFilePropertiesProvider;
import com.lasthopesoftware.bluewater.servers.library.items.media.files.properties.FilePropertiesProvider;
import com.lasthopesoftware.bluewater.shared.IoCommon;
import com.lasthopesoftware.bluewater.shared.MagicPropertyBuilder;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Map;
import java.util.concurrent.ExecutionException;

/**
 * Created by david on 7/24/15.
 */
public class MediaFileUriProvider extends AbstractFileUriProvider {

	public static final String mediaFileFoundEvent = MagicPropertyBuilder.buildMagicPropertyName(MediaFileUriProvider.class, "mediaFileFoundEvent");
	public static final String mediaFileFoundMediaId = MagicPropertyBuilder.buildMagicPropertyName(MediaFileUriProvider.class, "mediaFileFoundMediaId");
	public static final String mediaFileFoundFileKey = MagicPropertyBuilder.buildMagicPropertyName(MediaFileUriProvider.class, "mediaFileFoundFileKey");
	public static final String mediaFileFoundPath = MagicPropertyBuilder.buildMagicPropertyName(MediaFileUriProvider.class, "mediaFileFoundPath");

	private static final String audioIdKey = MediaStore.Audio.keyFor("audio_id");
	private static final String mediaDataQuery = MediaStore.Audio.Media.DATA + " LIKE '%' || ? || '%' ";
	private static final String[] mediaQueryProjection = { MediaStore.Audio.Media.DATA };

	private static final Logger logger = LoggerFactory.getLogger(MediaFileUriProvider.class);

	private final Context context;
	private final boolean isSilent;
	private final IConnectionProvider connectionProvider;

	public MediaFileUriProvider(Context context, IConnectionProvider connectionProvider, IFile file) {
		this (context, connectionProvider, file, false);
	}

	/**
	 *
	 * @param context the application context under which to operate
	 * @param file the file to provide a URI for
	 * @param isSilent if true, will not emit broadcast events when media files are found
	 */
	public MediaFileUriProvider(Context context, IConnectionProvider connectionProvider, IFile file, boolean isSilent) {
		super(file);

		this.context = context;
		this.isSilent = isSilent;
		this.connectionProvider = connectionProvider;
	}

	@Override
	public Uri getFileUri() throws IOException {
		final Cursor cursor = getMediaQueryCursor();
		if (cursor == null) return null;

		try {
			if (!cursor.moveToFirst()) return null;

			final String fileUriString = cursor.getString(cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATA));
			if (fileUriString == null || fileUriString.isEmpty()) return null;

			// The file object will produce a properly escaped File URI, as opposed to what is stored in the DB
			final java.io.File file = new java.io.File(fileUriString.replaceFirst(IoCommon.FileUriScheme + "://", ""));

			if (!file.exists()) return null;

			if (!isSilent) {
				final Intent broadcastIntent = new Intent(mediaFileFoundEvent);
				broadcastIntent.putExtra(mediaFileFoundPath, file.getPath());
				try {
					broadcastIntent.putExtra(mediaFileFoundMediaId, cursor.getInt(cursor.getColumnIndexOrThrow(audioIdKey)));
				} catch (IllegalArgumentException ie) {
					logger.info("Illegal column name.", ie);
				}
				broadcastIntent.putExtra(mediaFileFoundFileKey, getFile().getKey());
				LocalBroadcastManager.getInstance(context).sendBroadcast(broadcastIntent);
			}

			logger.info("Returning file URI from local disk.");
			return Uri.fromFile(file);
		} finally {
			cursor.close();
		}
	}

	public int getMediaId() throws IOException {
		final Cursor cursor = getMediaQueryCursor();
		if (cursor == null) return -1;

		try {
			if (cursor.moveToFirst())
				return cursor.getInt(cursor.getColumnIndexOrThrow(audioIdKey));
		} catch (IllegalArgumentException ie) {
			logger.info("Illegal column name.", ie);
		} finally {
			cursor.close();
		}

		return -1;
	}

	private Cursor getMediaQueryCursor() throws IOException {
		if (context == null)
			throw new NullPointerException("The file player's context cannot be null");

		if (ContextCompat.checkSelfPermission(context, Manifest.permission.READ_EXTERNAL_STORAGE) == PackageManager.PERMISSION_DENIED)
			return null;

		final CachedFilePropertiesProvider filePropertiesProvider = new CachedFilePropertiesProvider(connectionProvider, getFile().getKey());

		try {
			final Map<String, String> fileProperties = filePropertiesProvider.get();

			final String originalFilename = fileProperties.get(FilePropertiesProvider.FILENAME);
			if (originalFilename == null)
				throw new IOException("The filename property was not retrieved. A connection needs to be re-established.");

			final String filename = originalFilename.substring(originalFilename.lastIndexOf('\\') + 1, originalFilename.lastIndexOf('.'));

			final StringBuilder querySb = new StringBuilder(mediaDataQuery);
			appendAnd(querySb);

			final ArrayList<String> params = new ArrayList<>(5);
			params.add(filename);

			appendPropertyFilter(querySb, params, MediaStore.Audio.Media.ARTIST, fileProperties.get(FilePropertiesProvider.ARTIST));
			appendAnd(querySb);

			appendPropertyFilter(querySb, params, MediaStore.Audio.Media.ALBUM, fileProperties.get(FilePropertiesProvider.ALBUM));
			appendAnd(querySb);

			appendPropertyFilter(querySb, params, MediaStore.Audio.Media.TITLE, fileProperties.get(FilePropertiesProvider.NAME));
			appendAnd(querySb);

			appendPropertyFilter(querySb, params, MediaStore.Audio.Media.TRACK, fileProperties.get(FilePropertiesProvider.TRACK));

			return context.getContentResolver().query(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, mediaQueryProjection, querySb.toString(), params.toArray(new String[params.size()]), null);
		} catch (InterruptedException e) {
			return null;
		} catch (ExecutionException e) {
			if (e.getCause() instanceof FileNotFoundException)
				throw (FileNotFoundException)e.getCause();

			if (e.getCause() instanceof IOException)
				throw new IOException("The filename property was not retrieved. A connection needs to be re-established.", e.getCause());

			return null;
		}
	}

	private static StringBuilder appendPropertyFilter(final StringBuilder querySb, final ArrayList<String> params, final String key, final String value) {
		querySb.append(' ').append(key).append(' ');

		if (value != null) {
			querySb.append(" = ? ");
			params.add(value);
		} else {
			querySb.append(" IS NULL ");
		}

		return querySb;
	}

	private static StringBuilder appendAnd(final StringBuilder querySb) {
		return querySb.append(" AND ");
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/stored/library/items/files/external/MediaFileUriProvider.kt
package com.lasthopesoftware.bluewater.client.stored.library.items.files.external

import android.content.ContentResolver
import android.content.ContentUris
import android.database.Cursor
import android.net.Uri
import android.os.Build
import android.provider.MediaStore
import androidx.core.database.getLongOrNull
import com.lasthopesoftware.bluewater.client.browsing.files.ServiceFile
import com.lasthopesoftware.bluewater.client.browsing.files.properties.FilePropertyHelpers.fileNameParts
import com.lasthopesoftware.bluewater.client.browsing.files.properties.FilePropertyHelpers.localExternalRelativeFileDirectory
import com.lasthopesoftware.bluewater.client.browsing.files.properties.NormalizedFileProperties
import com.lasthopesoftware.bluewater.client.browsing.files.properties.ProvideLibraryFileProperties
import com.lasthopesoftware.bluewater.client.browsing.files.uri.ProvideFileUrisForLibrary
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.shared.android.permissions.CheckOsPermissions
import com.lasthopesoftware.bluewater.shared.lazyLogger
import com.lasthopesoftware.promises.extensions.preparePromise
import com.lasthopesoftware.resources.executors.ThreadPools
import com.lasthopesoftware.resources.uri.MediaCollections
import com.lasthopesoftware.resources.uri.resourceExists
import com.namehillsoftware.handoff.promises.Promise
import com.namehillsoftware.handoff.promises.response.PromisedResponse
import org.apache.commons.io.FilenameUtils
import java.io.IOException

private const val audioIdKey = MediaStore.Audio.Media._ID

private const val DISPLAY_NAME_MEDIA_COLLECTION_FILTER = "${MediaStore.Audio.Media.DISPLAY_NAME} LIKE ?"

private const val MEDIA_FILE_FOUND_LOG_MESSAGE = "Returning media file URI {} from local disk."

private val mediaQueryProjection = arrayOf(MediaStore.Audio.Media._ID, MediaStore.Audio.Media.DISPLAY_NAME)

private fun Cursor.takeFirstAudioFile() = use {
	it.takeIf { it.moveToFirst() }
		?.run { getLongOrNull(getColumnIndexOrThrow(audioIdKey)) }
		// The file object will produce a properly escaped file URI, as opposed to what is stored in the DB
		?.let { fileId -> ContentUris.withAppendedId(MediaCollections.ExternalAudio, fileId) }
}

interface MediaFileUriProvider : ProvideFileUrisForLibrary

abstract class PermissionsCheckingMediaFileUriProvider(
	private val filePropertiesProvider: ProvideLibraryFileProperties,
	private val externalStorageReadPermissionsArbitrator: CheckOsPermissions,
): MediaFileUriProvider, PromisedResponse<Map<String, String>, Uri?> {
	override fun promiseUri(libraryId: LibraryId, serviceFile: ServiceFile): Promise<Uri?> =
		if (externalStorageReadPermissionsArbitrator.run { !isReadPermissionGranted && !isReadMediaAudioPermissionGranted }) Promise.empty()
		else getMediaQueryUri(libraryId, serviceFile)

	private fun getMediaQueryUri(libraryId: LibraryId, serviceFile: ServiceFile): Promise<Uri?> =
		filePropertiesProvider
			.promiseFileProperties(libraryId, serviceFile)
			.eventually(this)
}

class DataFileUriProvider(
	filePropertiesProvider: ProvideLibraryFileProperties,
	externalStorageReadPermissionsArbitrator: CheckOsPermissions,
	private val contentResolver: ContentResolver,
) : PermissionsCheckingMediaFileUriProvider(filePropertiesProvider, externalStorageReadPermissionsArbitrator), PromisedResponse<Map<String, String>, Uri?> {

	override fun promiseResponse(fileProperties: Map<String, String>): Promise<Uri?> {
		val (_, baseFileName, _, postExtension) = fileProperties.fileNameParts
			?: throw IOException("The filename property was not retrieved. A connection needs to be re-established.")

		val fileNamePattern = FilenameUtils.concat(
				"%${fileProperties.localExternalRelativeFileDirectory}",
				if (postExtension.isEmpty()) "%$baseFileName%." else "%$postExtension%.")

		return ThreadPools.io.preparePromise {
			val maybeCursor = contentResolver.query(
				MediaCollections.ExternalAudio,
				mediaQueryProjection,
				DISPLAY_NAME_MEDIA_COLLECTION_FILTER,
				arrayOf(fileNamePattern),
				null
			)

			maybeCursor
				?.takeFirstAudioFile()
				?.takeIf(contentResolver::resourceExists)?.also {
					logger.info(MEDIA_FILE_FOUND_LOG_MESSAGE, it)
				}
		}
	}

	companion object {
        private val logger by lazyLogger<DataFileUriProvider>()
    }
}

class MetadataMediaFileUriProvider(
	filePropertiesProvider: ProvideLibraryFileProperties,
	externalStorageReadPermissionsArbitrator: CheckOsPermissions,
	private val contentResolver: ContentResolver,
) : PermissionsCheckingMediaFileUriProvider(filePropertiesProvider, externalStorageReadPermissionsArbitrator), PromisedResponse<Map<String, String>, Uri?> {

	override fun promiseResponse(fileProperties: Map<String, String>): Promise<Uri?> = ThreadPools.io.preparePromise {
		val selectionArgs = arrayOf(
			fileProperties[NormalizedFileProperties.Artist] ?: "",
			fileProperties[NormalizedFileProperties.AlbumArtist] ?: "",
			fileProperties[NormalizedFileProperties.Name] ?: "",
			fileProperties[NormalizedFileProperties.Album] ?: ""
		)

		val maybeCursor = contentResolver.query(
			MediaCollections.ExternalAudio,
			mediaQueryProjection,
			META_DATA_FILTER,
			selectionArgs,
			null
		)

		maybeCursor
			?.takeFirstAudioFile()
			?.takeIf(contentResolver::resourceExists)?.also {
				logger.info(MEDIA_FILE_FOUND_LOG_MESSAGE, it)
			}
	}

	companion object {
		private val logger by lazyLogger<MetadataMediaFileUriProvider>()

		private const val META_DATA_FILTER = """${MediaStore.Audio.Media.IS_PENDING} = 0
				AND COALESCE(${MediaStore.Audio.Media.ARTIST}, "") = ?
				AND COALESCE(${MediaStore.Audio.Media.ALBUM_ARTIST}, "") = ?
				AND COALESCE(${MediaStore.Audio.Media.TITLE}, "") = ?
				AND COALESCE(${MediaStore.Audio.AlbumColumns.ALBUM}, "") = ?"""
	}
}

class CompatibleMediaFileUriProvider(
	cachedFilePropertiesProvider: ProvideLibraryFileProperties,
	externalStorageReadPermissionsArbitrator: CheckOsPermissions,
	contentResolver: ContentResolver,
) : MediaFileUriProvider {
	private val inner: MediaFileUriProvider =
		if (Build.VERSION.SDK_INT >= 29) MetadataMediaFileUriProvider(cachedFilePropertiesProvider, externalStorageReadPermissionsArbitrator, contentResolver)
		else DataFileUriProvider(cachedFilePropertiesProvider, externalStorageReadPermissionsArbitrator, contentResolver)

	override fun promiseUri(libraryId: LibraryId, serviceFile: ServiceFile): Promise<Uri?> =
		inner.promiseUri(libraryId, serviceFile)
}
